File: Backend/app/schemas/product.py
from pydantic import BaseModel, HttpUrl, field_validator
from datetime import datetime
from typing import Optional, List, Generic, TypeVar
from enum import Enum


class ProductBase(BaseModel):
    url: str
    target_price: float
    check_frequency: Optional[int] = 24  # Default to 24 hours

    @field_validator('check_frequency')
    @classmethod
    def validate_check_frequency(cls, v):
        """Validate that check_frequency is one of the allowed values."""
        if v not in [6, 12, 24]:
            raise ValueError('check_frequency must be 6, 12, or 24 hours')
        return v


class ProductCreate(ProductBase):
    pass


class ProductUpdate(BaseModel):
    name: Optional[str] = None
    target_price: Optional[float] = None
    check_frequency: Optional[int] = None

    @field_validator('check_frequency')
    @classmethod
    def validate_check_frequency(cls, v):
        """Validate that check_frequency is one of the allowed values."""
        if v is not None and v not in [6, 12, 24]:
            raise ValueError('check_frequency must be 6, 12, or 24 hours')
        return v


class ProductResponse(BaseModel):
    id: int
    user_id: int
    name: str
    url: str
    image: Optional[str]
    current_price: float
    target_price: float
    last_checked: datetime
    created_at: datetime
    is_available: bool = True
    unavailable_since: Optional[datetime] = None
    check_frequency: int = 24

    class Config:
        from_attributes = True


class ProductScrapedData(BaseModel):
    name: str
    price: float
    image: Optional[str] = None


class SortOrder(str, Enum):
    """Sort order enum."""

    asc = "asc"
    desc = "desc"


class ProductSortBy(str, Enum):
    """Product sort field enum."""

    name = "name"
    price = "current_price"
    target_price = "target_price"
    created_at = "created_at"
    last_checked = "last_checked"


class PaginationMetadata(BaseModel):
    """Metadata for paginated responses."""

    page: int
    page_size: int
    total_items: int
    total_pages: int
    has_next: bool
    has_previous: bool


class PaginatedProductsResponse(BaseModel):
    """Paginated response for products."""

    items: List[ProductResponse]
    metadata: PaginationMetadata

File: Frontend/app/src/types/product.ts
export interface Product {
  id: number;
  user_id: number;
  name: string;
  url: string;
  image: string | null;
  current_price: number;
  target_price: number;
  last_checked: string;
  created_at: string;
  is_available: boolean;
  unavailable_since: string | null;
  check_frequency: 6 | 12 | 24;
}

export interface ProductCreate {
  url: string;
  target_price: number;
  check_frequency?: 6 | 12 | 24;
}

export interface ProductUpdate {
  name?: string;
  target_price?: number;
  check_frequency?: 6 | 12 | 24;
}

export interface PaginationMetadata {
  page: number;
  page_size: number;
  total_items: number;
  total_pages: number;
  has_next: boolean;
  has_previous: boolean;
}

export interface PaginatedProducts {
  items: Product[];
  metadata: PaginationMetadata;
}

// Price history interface for tracking product price changes over time
export interface PriceHistory {
  id: number;
  product_id: number;
  price: number;
  recorded_at: string;
}

// Price statistics interface for displaying aggregated price data
export interface PriceStats {
  current_price: number;
  lowest_price: number;
  highest_price: number;
  average_price: number;
  price_change_percentage: number | null;
  total_records: number;
}

export type SortBy = 'name' | 'current_price' | 'target_price' | 'created_at' | 'last_checked';
export type SortOrder = 'asc' | 'desc';

export interface ProductFilters {
  page?: number;
  page_size?: number;
  search?: string;
  sort_by?: SortBy;
  order?: SortOrder;
}
